### PR TITLE
feat: optional args for regexp_* UDFs

### LINKED ISSUE #10512: feat: enable optional UDF arguments in `regexp_*` functions
### Is your feature request related to a problem or challenge?

The same discussion / reasoning as in #10424.

The `inner` implementation for these UDFs changes based on the number of parameters passed, but the API of the exported `regex::expr_fn::*` does not enable that.

### Describe the solution you'd like

The same solution as implemented #10469, change the exported `expr_fn`s to take `Option<Expr>` for optional arguments. 

### Describe alternatives you've considered

Alternatives discussed in the prior issue:

- change expr_fn to take a `Vec` instead of explicit arguments
- maintain a different exported `expr_fn` for each set of non-null parameters

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Michael-J-Ward**
## Which issue does this PR close?

Closes #10512

## Rationale for this change

The inner implementations of these UDFs depend on the number of arguments passed. This exposes that to users of the UDFs.

## What changes are included in this PR?

for each of `regexp_match` regexp_like` and `regexp_replace`

- Changes / adds optional arguments to match inner impl
- Updates argument names to match inner impl

## Are these changes tested?

Yes. Each function already had a test case, and were updated to include variants for the optional arguments.

## Are there any user-facing changes?

Yes.

- `expr_fn::regexp_replace` signature has changed
- `expr_fn::regexp_like` signature has changed
- `expr_fn::regexp_match` signature has changed

### CODE REVIEW

**Comment by Michael-J-Ward** [comment]
Requesting review from @jayzhan211 and @jonahgao

File: datafusion/functions/src/regex/mod.rs [hunk 1]
```diff
@@ -28,12 +28,53 @@ make_udf_function!(
...
-    "returns a list of regular expression matches in a string. "
-),(
-    regexp_like,
-    input_arg1 input_arg2,
-    "Returns true if a has at least one match in a string,false otherwise."
-),(regexp_replace, arg1 arg2 arg3 arg4, "Replaces substrings in a string that match"));
+
+pub mod expr_fn {
+    #[doc = "returns a list of regular expression matches in a string. "]
+    #[doc = r" Return $name(arg)"]
```
Comment (jonahgao) [inline datafusion/functions/src/regex/mod.rs:34]: `#[doc = r" Return $name(arg)"]`  looks a bit weird, maybe we can delete it.

File: datafusion/functions/src/regex/mod.rs [hunk 2]
```diff
@@ -28,12 +28,53 @@ make_udf_function!(
...
+        super::regexp_like().call(args)
+    }
+
+    #[doc = "Replaces substrings in a string that match"]
+    #[doc = r" Return $name(arg)"]
+    pub fn regexp_replace(
+        string: datafusion_expr::Expr,
+        pattern: datafusion_expr::Expr,
+        replacement: datafusion_expr::Expr,
+        flags: Option<datafusion_expr::Expr>,
```
Comment (jonahgao) [inline datafusion/functions/src/regex/mod.rs:67]: Perhaps we can simplify the types of these parameters by using a use declaration.
```rust
pub mod expr_fn {
    use datafusion_expr::Expr;

    #[doc = "returns a list of regular expression matches in a string. "]
    pub fn regexp_match(values: Expr, regex: Expr, flags: Option<Expr>) -> Expr {
        let mut args = vec![values, regex];
        if let Some(flags) = flags {
            args.push(flags);
        };
        super::regexp_match().call(args)
    }
```

File: datafusion/functions/src/regex/mod.rs (hunk 1 above)
Comment (Michael-J-Ward) [reply datafusion/functions/src/regex/mod.rs]: Done

File: datafusion/functions/src/regex/mod.rs (hunk 2 above)
Comment (Michael-J-Ward) [reply datafusion/functions/src/regex/mod.rs]: Done

**Review by jonahgao — APPROVED** [review]
Looks good to me! Thank you @Michael-J-Ward

**Pull request merged**